File: dart_copy/lib/src/features/auth/data/auth_repository.dart
import 'package:drift/drift.dart';

import '../../../core/database/app_database.dart';
import '../domain/account.dart';
import 'account_dao.dart';
import 'password_hasher.dart';

/// Sign-up, sign-in and profile edits against the local accounts table.
///
/// Every method throws [AuthException] on a rejected attempt, so the UI has
/// one thing to catch and one message to show.
class AuthRepository {
  const AuthRepository(this._dao);

  static final RegExp _usernamePattern = RegExp(r'^[a-z0-9._]{3,24}$');

  final AccountDao _dao;

  /// Creates a local account and returns it signed in.
  Future<Account> signUp({
    required String username,
    required String password,
    required String displayName,
    String? farmName,
    String? email,
    DateTime? now,
  }) async {
    final handle = username.trim().toLowerCase();
    if (!_usernamePattern.hasMatch(handle)) {
      throw const AuthException(AuthFailure.invalidUsername);
    }
    if (!isPasswordAcceptable(password)) {
      throw const AuthException(AuthFailure.weakPassword);
    }
    if (await _dao.findByUsername(handle) != null) {
      throw const AuthException(AuthFailure.usernameTaken);
    }

    final digest = PasswordHasher.hash(password);
    final createdAt = now ?? DateTime.now();
    final name = displayName.trim().isEmpty ? handle : displayName.trim();

    final id = await _dao.insertAccount(
      AccountsCompanion.insert(
        username: handle,
        displayName: name,
        farmName: Value(_orNull(farmName)),
        email: Value(_orNull(email)),
        passwordHash: digest.hash,
        passwordSalt: digest.salt,
        hashIterations: digest.iterations,
        createdAt: createdAt,
        lastLoginAt: Value(createdAt),
      ),
    );

    return Account(
      id: id,
      username: handle,
      displayName: name,
      farmName: _orNull(farmName),
      email: _orNull(email),
      createdAt: createdAt,
      lastLoginAt: createdAt,
    );
  }

  Future<Account> signIn({
    required String username,
    required String password,
    DateTime? now,
  }) async {
    final row = await _dao.findByUsername(username.trim());
    if (row == null) {
      throw const AuthException(AuthFailure.unknownUser);
    }

    final matches = PasswordHasher.verify(
      password,
      hash: row.passwordHash,
      salt: row.passwordSalt,
      iterations: row.hashIterations,
    );
    if (!matches) {
      throw const AuthException(AuthFailure.wrongPassword);
    }

    final when = now ?? DateTime.now();
    await _dao.touchLastLogin(row.id, when);
    return _toDomain(row).copyWith(lastLoginAt: when);
  }

  Future<Account?> findById(int id) async {
    final row = await _dao.findById(id);
    return row == null ? null : _toDomain(row);
  }

  Stream<Account?> watchById(int id) =>
      _dao.watchById(id).map((row) => row == null ? null : _toDomain(row));

  /// True when at least one account exists, which decides whether the app
  /// opens on Sign in or on Sign up.
  Future<bool> hasAnyAccount() async => await _dao.countAll() > 0;

  Future<Account> updateProfile(
    Account account, {
    String? displayName,
    String? farmName,
    String? email,
  }) async {
    final updated = account.copyWith(
      displayName: displayName?.trim(),
      farmName: farmName?.trim(),
      email: email?.trim(),
    );

    await _dao.updateAccount(
      AccountsCompanion(
        displayName: Value(updated.displayName),
        farmName: Value(_orNull(updated.farmName)),
        email: Value(_orNull(updated.email)),
      ),
      id: account.id,
    );

    return updated;
  }

  Future<void> changePassword(
    Account account, {
    required String currentPassword,
    required String newPassword,
  }) async {
    final row = await _dao.findById(account.id);
    if (row == null) throw const AuthException(AuthFailure.unknownUser);

    final matches = PasswordHasher.verify(
      currentPassword,
      hash: row.passwordHash,
      salt: row.passwordSalt,
      iterations: row.hashIterations,
    );
    if (!matches) throw const AuthException(AuthFailure.wrongPassword);
    if (!isPasswordAcceptable(newPassword)) {
      throw const AuthException(AuthFailure.weakPassword);
    }

    final digest = PasswordHasher.hash(newPassword);
    await _dao.updateAccount(
      AccountsCompanion(
        passwordHash: Value(digest.hash),
        passwordSalt: Value(digest.salt),
        hashIterations: Value(digest.iterations),
      ),
      id: account.id,
    );
  }

  /// At least 8 characters with a letter and a digit. Modest, but it is the
  /// only barrier on an offline device and the beekeeper types it in a shed.
  static bool isPasswordAcceptable(String password) {
    return password.length >= 8 &&
        password.contains(RegExp('[A-Za-z]')) &&
        password.contains(RegExp('[0-9]'));
  }

  static String? _orNull(String? value) {
    final trimmed = value?.trim();
    return trimmed == null || trimmed.isEmpty ? null : trimmed;
  }

  Account _toDomain(AccountRow row) => Account(
        id: row.id,
        username: row.username,
        displayName: row.displayName,
        farmName: row.farmName,
        email: row.email,
        createdAt: row.createdAt,
        lastLoginAt: row.lastLoginAt,
      );
}

File: react_native/src/features/auth/data/authRepository.ts
import { count, eq } from 'drizzle-orm';

import type { AppDatabase } from '@/core/database/client';
import { accounts, type AccountRow } from '@/core/database/schema';

import { AuthError, type Account } from '../domain/account';
import { hashPassword, verifyPassword } from './passwordHasher';

const USERNAME_PATTERN = /^[a-z0-9._]{3,24}$/;

export interface SignUpInput {
  username: string;
  password: string;
  displayName: string;
  farmName?: string | null;
  email?: string | null;
  now?: Date;
}

export interface AuthRepository {
  /** Creates a local account and returns it signed in. */
  signUp(input: SignUpInput): Promise<Account>;
  signIn(username: string, password: string, now?: Date): Promise<Account>;
  findById(id: number): Promise<Account | null>;
  /**
   * True when at least one account exists, which decides whether the app
   * opens on Sign in or on Sign up.
   */
  hasAnyAccount(): Promise<boolean>;
  updateProfile(
    account: Account,
    changes: { displayName?: string; farmName?: string; email?: string },
  ): Promise<Account>;
  changePassword(account: Account, currentPassword: string, newPassword: string): Promise<void>;
}

/**
 * At least 8 characters with a letter and a digit. Modest, but it is the
 * only barrier on an offline device and the beekeeper types it in a shed.
 */
export function isPasswordAcceptable(password: string): boolean {
  return password.length >= 8 && /[A-Za-z]/.test(password) && /[0-9]/.test(password);
}

const orNull = (value: string | null | undefined): string | null => {
  const trimmed = value?.trim();
  return trimmed == null || trimmed.length === 0 ? null : trimmed;
};

export function accountFromRow(row: AccountRow): Account {
  return {
    id: row.id,
    username: row.username,
    displayName: row.displayName,
    farmName: row.farmName,
    email: row.email,
    createdAt: row.createdAt,
    lastLoginAt: row.lastLoginAt,
  };
}

/**
 * Sign-up, sign-in and profile edits against the local accounts table.
 *
 * Every method throws [AuthError] on a rejected attempt, so the UI has one
 * thing to catch and one message to show.
 */
export class DrizzleAuthRepository implements AuthRepository {
  constructor(private readonly db: AppDatabase) {}

  private async findRowByUsername(username: string): Promise<AccountRow | null> {
    const [row] = await this.db
      .select()
      .from(accounts)
      .where(eq(accounts.username, username.toLowerCase()))
      .limit(1);
    return row ?? null;
  }

  private async findRowById(id: number): Promise<AccountRow | null> {
    const [row] = await this.db.select().from(accounts).where(eq(accounts.id, id)).limit(1);
    return row ?? null;
  }

  async signUp(input: SignUpInput): Promise<Account> {
    const handle = input.username.trim().toLowerCase();
    if (!USERNAME_PATTERN.test(handle)) throw new AuthError('invalidUsername');
    if (!isPasswordAcceptable(input.password)) throw new AuthError('weakPassword');
    if ((await this.findRowByUsername(handle)) != null) throw new AuthError('usernameTaken');

    const digest = await hashPassword(input.password);
    const createdAt = input.now ?? new Date();
    const trimmedName = input.displayName.trim();
    const name = trimmedName.length === 0 ? handle : trimmedName;

    const [inserted] = await this.db
      .insert(accounts)
      .values({
        username: handle,
        displayName: name,
        farmName: orNull(input.farmName),
        email: orNull(input.email),
        passwordHash: digest.hash,
        passwordSalt: digest.salt,
        hashIterations: digest.iterations,
        createdAt,
        lastLoginAt: createdAt,
      })
      .returning({ id: accounts.id });

    return {
      id: inserted.id,
      username: handle,
      displayName: name,
      farmName: orNull(input.farmName),
      email: orNull(input.email),
      createdAt,
      lastLoginAt: createdAt,
    };
  }

  async signIn(username: string, password: string, now: Date = new Date()): Promise<Account> {
    const row = await this.findRowByUsername(username.trim());
    if (row == null) throw new AuthError('unknownUser');

    const matches = await verifyPassword(password, {
      hash: row.passwordHash,
      salt: row.passwordSalt,
      iterations: row.hashIterations,
    });
    if (!matches) throw new AuthError('wrongPassword');

    await this.db.update(accounts).set({ lastLoginAt: now }).where(eq(accounts.id, row.id));
    return { ...accountFromRow(row), lastLoginAt: now };
  }

  async findById(id: number): Promise<Account | null> {
    const row = await this.findRowById(id);
    return row ? accountFromRow(row) : null;
  }

  async hasAnyAccount(): Promise<boolean> {
    const [row] = await this.db.select({ value: count() }).from(accounts);
    return (row?.value ?? 0) > 0;
  }

  async updateProfile(
    account: Account,
    changes: { displayName?: string; farmName?: string; email?: string },
  ): Promise<Account> {
    const updated: Account = {
      ...account,
      displayName: changes.displayName?.trim() ?? account.displayName,
      farmName: changes.farmName?.trim() ?? account.farmName,
      email: changes.email?.trim() ?? account.email,
    };

    await this.db
      .update(accounts)
      .set({
        displayName: updated.displayName,
        farmName: orNull(updated.farmName),
        email: orNull(updated.email),
      })
      .where(eq(accounts.id, account.id));

    return updated;
  }

  async changePassword(
    account: Account,
    currentPassword: string,
    newPassword: string,
  ): Promise<void> {
    const row = await this.findRowById(account.id);
    if (row == null) throw new AuthError('unknownUser');

    const matches = await verifyPassword(currentPassword, {
      hash: row.passwordHash,
      salt: row.passwordSalt,
      iterations: row.hashIterations,
    });
    if (!matches) throw new AuthError('wrongPassword');
    if (!isPasswordAcceptable(newPassword)) throw new AuthError('weakPassword');

    const digest = await hashPassword(newPassword);
    await this.db
      .update(accounts)
      .set({
        passwordHash: digest.hash,
        passwordSalt: digest.salt,
        hashIterations: digest.iterations,
      })
      .where(eq(accounts.id, account.id));
  }
}
